File: codepropertygraph/src/main/java/io/shiftleft/codepropertygraph/cpgloading/ZipArchive.java
package io.shiftleft.codepropertygraph.cpgloading;

import static java.nio.file.Files.walkFileTree;

import java.io.Closeable;
import java.io.IOException;
import java.nio.file.FileSystem;
import java.nio.file.FileSystems;
import java.nio.file.FileVisitResult;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.SimpleFileVisitor;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.ArrayList;
import java.util.Collection;

public class ZipArchive implements Closeable {

  private final FileSystem zipFileSystem;

  ZipArchive(String inputFile) throws IOException {
    this.zipFileSystem = FileSystems.newFileSystem(Paths.get(inputFile), null);
  }

  public Collection<Path> getFileEntries() throws IOException {
    final Path root = zipFileSystem.getRootDirectories().iterator().next();
    ArrayList<Path> entries = new ArrayList<>();

    walkFileTree(root, new SimpleFileVisitor<Path>() {
      @Override
      public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
        if (attrs.isRegularFile()) {
          entries.add(file);
        }
        return FileVisitResult.CONTINUE;
      }
    });

    return entries;
  }

  @Override
  public void close() throws IOException {
    zipFileSystem.close();
  }
}

File: codepropertygraph/src/main/scala/io/shiftleft/codepropertygraph/cpgloading/ZipArchive.scala
package io.shiftleft.codepropertygraph.cpgloading

import java.io.Closeable
import java.nio.file.attribute.BasicFileAttributes
import java.nio.file.{FileSystem, FileSystems, FileVisitOption, FileVisitResult, Files, Path, Paths, SimpleFileVisitor}
import java.util.Collection as JCollection
import scala.collection.immutable.ArraySeq
import scala.collection.mutable.ArrayBuffer
import scala.jdk.CollectionConverters.*

class ZipArchive(inputFile: String) extends Closeable {
  private val zipFileSystem: FileSystem = FileSystems.newFileSystem(Paths.get(inputFile), null: ClassLoader)

  private def root: Path = zipFileSystem.getRootDirectories.iterator.next

  private def walk(rootPath: Path): Seq[Path] = {
    val entries = ArraySeq.newBuilder[Path]()
    Files.walkFileTree(
      rootPath,
      new SimpleFileVisitor[Path]() {
        override def visitFile(file: Path, attrs: BasicFileAttributes): FileVisitResult = {
          if (attrs.isRegularFile)
            entries += file
          FileVisitResult.CONTINUE
        }
      }
    )
    entries.result()
  }

  def entries: Seq[Path] = walk(root)

  def getFileEntries: JCollection[Path] = entries.asJava

  override def close(): Unit = zipFileSystem.close()
}
